lib/holepuncher.js

```diff
@@ -0,0 +1,288 @@
+const utp = require('utp-native')
+const os = require('os')
+const Timer = require('./timer')
+
+const BIRTHDAY_SOCKETS = 256
+const HOLEPUNCH = Buffer.from([0])
+const HOLEPUNCH_TTL = 5
+const DEFAULT_TTL = 64
+
+// const NAT_OPEN = Symbol.for('NAT_OPEN')
+const NAT_UNKNOWN = Symbol.for('NAT_UNKNOWN')
+const NAT_PORT_RANDOMIZED = Symbol.for('NAT_PORT_RANDOMIZED')
+
+module.exports = class Holepuncher {
+  constructor (address) {
+    const self = this
+
+    this.address = address
+    this.socket = null
+
+    this.localNetwork = null
+    this.remoteNetwork = null
+    this.localFriendly = false
+    this.remoteFriendly = false
+    this.lan = false
+
+    this._onmessage = onmessage
+    this._allSockets = []
+    this._holepunchTimer = null
+    this._lanTimer = null
+    this._stopped = false
+    this._isClient = false
+
+    // async states
+    this._openingSessions = null
+    this._holepunching = null
+    this._resolveConnection = null
+    this._onconnection = onconnection
+
+    this.connected = new Promise((resolve) => { this._resolveConnection = resolve })
+
+    function onconnection (rawSocket) {
+      if (self._resolveConnection === null) {
+        self._stop(null)
+        rawSocket.on('error', noop)
+        rawSocket.destroy()
+        return
+      }
+
+      const socket = this
```

Comment:
is that the same as `self.socket`? if so maybe this could be taken out of the constructor and made a private method?

Reply:
if would easy if that was true wouldn't it? 🤣 This is just my crazy way of reusing the event handler everywhere and find out which socket fired it using a `this` ref in there.

Reply:
hahah I see - I'd say leave for now and then we re-assess after full test cov